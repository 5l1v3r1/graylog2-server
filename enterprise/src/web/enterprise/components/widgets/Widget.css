:local(.widgetActionDropdownCaret) {
    cursor: pointer;
    opacity: 0.3;
    font-size: 14px !important;
}
