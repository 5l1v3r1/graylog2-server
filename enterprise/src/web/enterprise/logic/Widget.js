import uuid from 'uuid/v4';

import { PluginStore } from 'graylog-web-plugin/plugin';

import { TIMESTAMP_FIELD } from 'enterprise/Constants';
import pivotForField from './searchtypes/aggregation/PivotGenerator';
import AggregationWidget from './aggregationbuilder/AggregationWidget';
import AggregationWidgetConfig from './aggregationbuilder/AggregationWidgetConfig';
import Series from './aggregationbuilder/Series';
import FieldType from './fieldtypes/FieldType';

const widgetsKey = 'enterpriseWidgets';

export function widgetDefinition(type) {
  return PluginStore.exports(widgetsKey)
    .find(widget => widget.type.toLocaleUpperCase() === type.toLocaleUpperCase());
}

export const resultHistogram = (id = uuid()) => AggregationWidget.builder()
  .id(id)
  .config(
    AggregationWidgetConfig.builder()
      .columnPivots([])
      .rowPivots([
        pivotForField(TIMESTAMP_FIELD, new FieldType('date', [], [])),
      ])
      .series([
        Series.forFunction('count()'),
      ])
      .sort([])
      .visualization('bar')
      .rollup(true)
      .build(),
  )
  .build();
