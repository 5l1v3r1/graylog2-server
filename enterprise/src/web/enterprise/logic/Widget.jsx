import { PluginStore } from 'graylog-web-plugin/plugin';

const widgetsKey = 'enterpriseWidgets';

export function widgetDefinition(type) {
  return PluginStore.exports(widgetsKey)
    .find(widget => widget.type.toLocaleUpperCase() === type.toLocaleUpperCase())
}

export const messageList = (id, timeRange, fields) => {
  return {
    id: id,
    type: 'messages',
    title: 'Messages',
    config: {
      fields: fields,
      showMessageRow: true,
    },
    computationTimeRange: timeRange,
  };
};

export const resultHistogram = (id, timeRange = {}) => {
  return {
    id: id,
    type: 'SEARCH_RESULT_CHART2',
    title: 'Histogram',
    computationTimeRange: timeRange,
    config: {
      timerange: timeRange,
    },
  };
};

export const searchSideBar = () => {
  return {
    type: 'SEARCH_SIDEBAR',
    title: 'Search Result',
    config: {},
    data: 'messages',
  };
};

export const dataTable = (id) => {
  return {
    id,
    type: 'DATATABLE',
    title: 'Results',
    config: {
      fields: ['action', 'controller', 'count'],
      data: [
        { action: 'index', controller: 'PostsController', count: 19016 },
        { controller: 'UsersController', count: 1376 },
        { action: 'login', controller: 'LoginController', count: 5156 },
        { action: 'show', controller: 'PostsController', count: 5903 },
        { action: 'edit', controller: 'PostsController', count: 922 },
      ],
    },
  };
};

const createWidgetDefinitions = (timeRange, fields) => {
  return {
    data: {},
    widgets: {
      messages: messageList(timeRange, fields),
      histogram: resultHistogram(timeRange),
      searchSideBar: searchSideBar(),
    },
    positions: {
      searchSideBar: { row: 1, col: 1, width: 1, height: 3.2 },
      histogram: { row: 1, col: 2, width: 4, height: 1.2 },
      messages: { row: 255, col: 2, width: 4, height: 3 },
    },
  };
};

