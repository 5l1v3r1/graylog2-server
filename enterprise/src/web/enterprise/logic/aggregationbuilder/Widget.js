import Immutable from 'immutable';

export default class Widget {
  constructor(id, type, config) {
    this._value = { id, type, config };
  }

  get id() {
    return this._value.id;
  }

  get type() {
    return this._value.type;
  }

  get config() {
    return this._value.config;
  }

  duplicate(newId) {
    return new Widget(newId, this.type, this.config);
  }

  toBuilder() {
    const { id, type, config } = this._value;
    return new Builder(Immutable.Map({ id, type, config }));
  }

  toJSON() {
    const { id, type, config } = this._value;

    return { id, type: type.toLocaleLowerCase(), config };
  }

  static fromJSON(value) {
    const { id, type, config } = value;
    return new Widget(id, type, Object.assign({}, config));
  }
}

class Builder {
  constructor(value = Immutable.Map()) {
    this.value = value;
  }

  id(value) {
    return new Builder(this.value.set('id', value));
  }

  type(value) {
    return new Builder(this.value.set('type', value));
  }

  config(value) {
    return new Builder(this.value.set('config', value));
  }

  build() {
    const { id, type, config } = this.value.toObject();
    return new Widget(id, type, config);
  }
}
