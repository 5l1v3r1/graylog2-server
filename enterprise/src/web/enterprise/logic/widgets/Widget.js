// @flow
import { Map } from 'immutable';
import uuid from 'uuid/v4';

type State = {
  id: string,
  type: string,
  config: any,
  filter: string;
};

class Widget {
  _value: State;
  constructor(id: string, type: string, config: any, filter: string) {
    this._value = { id, type, config, filter: filter === null ? undefined : filter };
  }

  static __registrations = {};

  get id() : string {
    return this._value.id;
  }

  get type() : string {
    return this._value.type;
  }

  get config() {
    return this._value.config;
  }

  get filter() : string {
    return this._value.filter;
  }

  duplicate(newId: string) : Widget {
    return this.toBuilder().id(newId).build();
  }

  toBuilder() : Builder {
    const { id, type, config, filter } = this._value;
    // eslint-disable-next-line no-use-before-define
    return new Builder(Map({ id, type, config, filter }));
  }

  toJSON() {
    const { id, type, config, filter } = this._value;

    return { id, type: type.toLocaleLowerCase(), config, filter };
  }

  static fromJSON(value: State) : Widget {
    const { id, type, config, filter } = value;
    const implementingClass = Widget.__registrations[type.toLocaleLowerCase()];

    if (implementingClass) {
      return implementingClass.fromJSON(value);
    }

    return new Widget(id, type, config, filter);
  }

  static registerSubtype(type: string, implementingClass: typeof Widget) {
    this.__registrations[type.toLocaleLowerCase()] = implementingClass;
  }
}

class Builder {
  value: Map<string, any>;
  constructor(value = Map()) {
    this.value = value;
  }

  id(value: string) {
    this.value = this.value.set('id', value);
    return this;
  }

  newId() {
    return this.id(uuid());
  }

  type(value: string) {
    this.value = this.value.set('type', value);
    return this;
  }

  config(value) {
    this.value = this.value.set('config', value);
    return this;
  }

  filter(value: string) {
    this.value = this.value.set('filter', value);
    return this;
  }

  build() : Widget {
    const { id, type, config, filter } = this.value.toObject();
    return new Widget(id, type, config, filter);
  }
}

Widget.Builder = Builder;

export default Widget;
