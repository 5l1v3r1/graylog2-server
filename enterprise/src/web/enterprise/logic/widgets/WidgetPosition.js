export default class WidgetPosition {
  constructor(col, row, height, width) {
    this._value = { col, row, height, width };
  }

  static fromJSON(value) {
    const { col, row, height, width } = value;
    const newCol = col === 'Infinity' ? Infinity : col;
    const newRow = row === 'Infinity' ? Infinity : row;
    const newHeight = height === 'Infinity' ? Infinity : height;
    const newWidth = width === 'Infinity' ? Infinity : width;

    return new WidgetPosition(newCol, newRow, newHeight, newWidth);
  }

  get col() {
    return this._value.col;
  }

  get row() {
    return this._value.row;
  }

  get height() {
    return this._value.height;
  }

  get width() {
    return this._value.width;
  }

  toJSON() {
    const { col, row, height, width } = this._value;

    const newCol = col === Infinity ? 'Infinity' : col;
    const newRow = row === Infinity ? 'Infinity' : row;
    const newHeight = height === Infinity ? 'Infinity' : height;
    const newWidth = width === Infinity ? 'Infinity' : width;

    return {
      col: newCol,
      row: newRow,
      height: newHeight,
      width: newWidth,
    };
  }
}
