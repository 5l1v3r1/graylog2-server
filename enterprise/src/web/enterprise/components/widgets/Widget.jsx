import React from 'react';
import PropTypes from 'prop-types';
import { MenuItem } from 'react-bootstrap';

import connect from 'stores/connect';
import UserNotification from 'util/UserNotification';
import { widgetDefinition } from 'enterprise/logic/Widget';
import { WidgetActions } from 'enterprise/stores/WidgetStore';
import { TitlesActions } from 'enterprise/stores/TitlesStore';
import { DashboardWidgetsActions } from 'enterprise/stores/DashboardWidgetsStore';

import WidgetFrame from './WidgetFrame';
import WidgetHeader from './WidgetHeader';
import WidgetFilterMenu from './WidgetFilterMenu';
import WidgetActionDropdown from './WidgetActionDropdown';
import WidgetHorizontalStretch from './WidgetHorizontalStretch';
import MeasureDimensions from './MeasureDimensions';

import styles from './Widget.css';
import EditWidgetFrame from './EditWidgetFrame';
import { ViewMetadataStore } from '../../stores/ViewMetadataStore';
import LoadingWidget from './LoadingWidget';
import ErrorWidget from './ErrorWidget';
import { WidgetErrorsMap } from './WidgetPropTypes';

class Widget extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    widget: PropTypes.shape({
      computationTimeRange: PropTypes.object,
      config: PropTypes.object.isRequired,
    }).isRequired,
    data: PropTypes.any,
    errors: WidgetErrorsMap,
    height: PropTypes.number,
    width: PropTypes.number,
    fields: PropTypes.any.isRequired,
    onSizeChange: PropTypes.func.isRequired,
    onPositionsChange: PropTypes.func.isRequired,
    title: PropTypes.string.isRequired,
    position: PropTypes.object.isRequired,
    view: PropTypes.shape({
      activeQuery: PropTypes.string,
    }).isRequired,
  };

  static defaultProps = {
    height: 1,
    width: 1,
    data: undefined,
    errors: undefined,
  };

  static _visualizationForType(type) {
    return widgetDefinition(type).visualizationComponent;
  }

  static _editComponentForType(type) {
    return widgetDefinition(type).editComponent;
  }

  constructor(props) {
    super(props);
    this.state = {
      editing: false,
    };
  }

  _onDelete = (widget) => {
    // eslint-disable-next-line no-alert
    if (window.confirm(`Are you sure you want to remove the widget "${this.props.title}"?`)) {
      WidgetActions.remove(widget.id);
    }
  };

  _onDuplicate = (widgetId) => {
    WidgetActions.duplicate(widgetId);
  };

  _onToggleEdit = () => {
    this.setState(state => ({ editing: !state.editing }));
  };

  _onAddToOverview = (queryId, widgetId) => {
    DashboardWidgetsActions.addToDashboard(queryId, widgetId)
      .then(
        () => UserNotification.success('Added widget to dashboard.', 'Success!'),
        e => UserNotification.error(`Failed adding widget to dashboard: ${e}`, 'Error!'),
      );
  };

  _onWidgetConfigChange = (widgetId, config) => {
    WidgetActions.updateConfig(widgetId, config);
  };

  visualize = () => {
    const { data, errors } = this.props;
    if (errors && errors.length > 0) {
      return <ErrorWidget errors={errors} />;
    }
    if (data) {
      const { editing } = this.state;
      const { id, widget, height, width, fields } = this.props;
      const { config, computationTimeRange, filter } = widget;
      const VisComponent = Widget._visualizationForType(widget.type);
      return (<VisComponent id={id}
                    editing={editing}
                    title={widget.title}
                    config={config}
                    data={data}
                    fields={fields}
                    height={height}
                    width={width}
                    filter={filter}
                    onFinishEditing={this._onToggleEdit}
                    computationTimeRange={computationTimeRange} />);
    }
    return <LoadingWidget />;
  };
  render() {
    const { id, widget, fields, onSizeChange, title } = this.props;
    const { editing } = this.state;
    const { config, filter } = widget;
    const { activeQuery } = this.props.view;
    const visualization = this.visualize();
    const widgetActionDropdownCaret = <i className={`fa fa-chevron-down ${styles.widgetActionDropdownCaret} ${styles.tonedDown}`} />;
    if (editing) {
      let editWidgetFrameContent = null;
      const EditComponent = Widget._editComponentForType(widget.type);
      return (
        <EditWidgetFrame widgetId={id}>
          <span ref={(elem) => { editWidgetFrameContent = elem; }}>
            <MeasureDimensions>
              <WidgetHeader title={title}
                            onRename={newTitle => TitlesActions.set('widget', id, newTitle)}
                            editing={editing}>
                <WidgetFilterMenu onChange={newFilter => WidgetActions.filter(id, newFilter)} value={filter}>
                  <i className={`fa fa-filter ${styles.widgetActionDropdownCaret} ${filter ? styles.filterSet : styles.filterNotSet}`} />
                </WidgetFilterMenu>
                {' '}
                <WidgetActionDropdown element={widgetActionDropdownCaret} container={() => editWidgetFrameContent}>
                  <MenuItem onSelect={this._onToggleEdit}>Finish Editing</MenuItem>
                </WidgetActionDropdown>
              </WidgetHeader>
              <EditComponent config={config}
                             fields={fields}
                             id={id}
                             onChange={newWidgetConfig => this._onWidgetConfigChange(id, newWidgetConfig)}>
                {visualization}
              </EditComponent>
            </MeasureDimensions>
          </span>
        </EditWidgetFrame>
      );
    }
    let container = null;
    return (
      <WidgetFrame widgetId={id} onSizeChange={onSizeChange}>
        <span ref={(elem) => { container = elem; }}>
          <MeasureDimensions>
            <WidgetHeader title={title}
                          onRename={newTitle => TitlesActions.set('widget', id, newTitle)}
                          editing={editing}>
              <WidgetHorizontalStretch widgetId={widget.id}
                                       widgetType={widget.type}
                                       onStretch={this.props.onPositionsChange}
                                       position={this.props.position} />
              {' '}
              <WidgetFilterMenu onChange={newFilter => WidgetActions.filter(id, newFilter)} value={filter}>
                <i className={`fa fa-filter ${styles.widgetActionDropdownCaret} ${filter ? styles.filterSet : styles.filterNotSet}`} />
              </WidgetFilterMenu>
              {' '}
              <WidgetActionDropdown element={widgetActionDropdownCaret} container={() => container}>
                <MenuItem onSelect={this._onToggleEdit}>Edit</MenuItem>
                <MenuItem onSelect={() => this._onDuplicate(id)}>Duplicate</MenuItem>
                <MenuItem divider />
                <MenuItem onSelect={() => this._onAddToOverview(activeQuery, id)}>Add to overview</MenuItem>
                <MenuItem divider />
                <MenuItem onSelect={() => this._onDelete(widget)}>Delete</MenuItem>
              </WidgetActionDropdown>
            </WidgetHeader>
            {visualization}
          </MeasureDimensions>
        </span>
      </WidgetFrame>
    );
  }
}

export default connect(Widget, { view: ViewMetadataStore });
