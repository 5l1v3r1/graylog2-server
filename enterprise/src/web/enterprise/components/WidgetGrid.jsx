import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { Row } from 'react-bootstrap';
import _ from 'lodash';

import connect from 'stores/connect';
import CustomPropTypes from 'enterprise/components/CustomPropTypes';
import style from 'pages/ShowDashboardPage.css';
import ReactGridContainer from 'components/common/ReactGridContainer';
import { widgetDefinition } from 'enterprise/logic/Widget';
import Widget from './widgets/Widget';
import { PositionsMap, WidgetsMap, WidgetDataMap } from './widgets/WidgetPropTypes';
import { TitlesStore } from '../stores/TitlesStore';
import WidgetPosition from '../logic/widgets/WidgetPosition';

const defaultTitleGenerator = w => `Unnamed ${w.type.replace('_', ' ').split(' ').map(_.capitalize).join(' ')}`;

class WidgetGrid extends React.Component {
  static _defaultDimensions(type) {
    const widgetDef = widgetDefinition(type);
    return new WidgetPosition(1, 1, widgetDef.defaultHeight, widgetDef.defaultWidth);
  }

  static _defaultTitle(widget) {
    const widgetDef = widgetDefinition(widget.type);
    return (widgetDef.titleGenerator || defaultTitleGenerator)(widget);
  }

  static propTypes = {
    allFields: CustomPropTypes.FieldListType.isRequired,
    data: WidgetDataMap.isRequired,
    fields: CustomPropTypes.FieldListType.isRequired,
    locked: PropTypes.bool,
    onPositionsChange: PropTypes.func.isRequired,
    positions: PositionsMap,
    staticWidgets: PropTypes.arrayOf(PropTypes.node),
    titles: ImmutablePropTypes.map.isRequired,
    widgets: WidgetsMap.isRequired,
  };

  static defaultProps = {
    locked: true,
    staticWidgets: [],
    positions: {},
  };

  state = {
    widgetDimensions: {},
  };

  _onWidgetSizeChange = (widgetId, dimensions) => {
    const widgetDimensions = (this.state && this.state.widgetDimensions) || {};
    widgetDimensions[widgetId] = dimensions;
    this.setState({ widgetDimensions: widgetDimensions });
  };

  _renderWidgets = (widgets, positions, data) => {
    const returnedWidgets = { positions: {}, widgets: [] };

    if (!widgets || _.isEmpty(widgets) || !data) {
      return returnedWidgets;
    }

    const onPositionsChange = (newPosition) => {
      const newPositions = Object.keys(positions).map((id) => {
        const { col, row, height, width } = positions[id]._value;
        return { id: id, col: col, row: row, height: height, width: width };
      });
      newPositions.push(newPosition);
      this.props.onPositionsChange(newPositions);
    };

    Object.keys(widgets).forEach((widgetId) => {
      const widget = widgets[widgetId];
      const dataKey = widget.data || widgetId;
      const widgetData = data[dataKey];

      returnedWidgets.positions[widgetId] = positions[widgetId] || WidgetGrid._defaultDimensions(widget.type);

      const { height, width } = (this.state && this.state.widgetDimensions[widgetId]) || {};

      const widgetTitle = this.props.titles.getIn(['widget', widget.id], WidgetGrid._defaultTitle(widget));

      returnedWidgets.widgets.push(
        <div key={widget.id} className={style.widgetContainer}>
          <Widget key={widgetId}
                  id={widgetId}
                  widget={widget}
                  data={widgetData}
                  height={height}
                  position={returnedWidgets.positions[widgetId]}
                  width={width}
                  allFields={this.props.allFields}
                  fields={this.props.fields}
                  onPositionsChange={onPositionsChange}
                  onSizeChange={this._onWidgetSizeChange}
                  title={widgetTitle} />
        </div>,
      );
    });

    return returnedWidgets;
  };

  render() {
    const { staticWidgets } = this.props;
    const { widgets, positions } = this._renderWidgets(this.props.widgets, this.props.positions, this.props.data);
    const grid = widgets && widgets.length > 0 ? (
      <ReactGridContainer animate={false}
                          locked={this.props.locked}
                          measureBeforeMount
                          onPositionsChange={this.props.onPositionsChange}
                          positions={positions}
                          useDragHandle=".widget-drag-handle">
        {widgets}
      </ReactGridContainer>
    ) : null;
    return (
      <Row>
        <div className="dashboard" style={{ marginLeft: '-20px' }}>
          {grid}
          {staticWidgets}
        </div>
      </Row>
    );
  }
}

export default connect(WidgetGrid, { titles: TitlesStore });
