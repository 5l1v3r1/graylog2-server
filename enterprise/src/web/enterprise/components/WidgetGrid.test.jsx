import React from 'react';
import renderer from 'react-test-renderer';
import Immutable from 'immutable';
import { mount } from 'enzyme';

import mockComponent from 'helpers/mocking/MockComponent';
import WidgetPosition from 'enterprise/logic/widgets/WidgetPosition';
import WidgetGrid from './WidgetGrid';
import Widget from './widgets/Widget';

jest.mock('./widgets/Widget', () => mockComponent('Widget'));
jest.mock('components/common/ReactGridContainer', () => mockComponent('ReactGridContainer'));
jest.mock('graylog-web-plugin/plugin', () => ({
  PluginStore: {
    exports: () => ([
      {
        type: 'dummy',
        displayName: 'Some Dummy Visualization',
        defaultHeight: 5,
        defaultWidth: 6,
      },
    ]),
  },
}));

describe('<WidgetGrid />', () => {
  it('should render with empty props', () => {
    const wrapper = renderer.create(<WidgetGrid />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with widgets passed', () => {
    const widgets = {
      widget1: { type: 'dummy', id: 'widget1' },
    };
    const positions = {
      widget1: new WidgetPosition(1, 1, 1, 1),
    };
    const data = {
      widget1: [],
    };

    const titles = Immutable.Map({
      widget1: 'A dummy widget',
    });
    const wrapper = mount(<WidgetGrid widgets={widgets} positions={positions} data={data} titles={titles} />);
    expect(wrapper.find(Widget)).toHaveLength(1);
  });

  it('should render widget even if widget has no data', () => {
    const widgets = {
      widget1: { type: 'dummy', id: 'widget1' },
    };
    const positions = {
      widget1: new WidgetPosition(1, 1, 1, 1),
    };
    const data = {
    };

    const titles = Immutable.Map({
      widget1: 'A dummy widget',
    });
    const wrapper = mount(<WidgetGrid widgets={widgets} positions={positions} data={data} titles={titles} />);
    expect(wrapper.find(Widget)).toHaveLength(1);
  });
});