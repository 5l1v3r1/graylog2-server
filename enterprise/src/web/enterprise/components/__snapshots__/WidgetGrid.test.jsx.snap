// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WidgetGrid /> should render with empty props 1`] = `
<div
  className="row"
>
  <div
    className="dashboard"
    style={
      Object {
        "marginLeft": "-20px",
      }
    }
  />
</div>
`;
